pywren/storage/storage.py

```diff
@@ -0,0 +1,140 @@
+import os
+import sys
+import json
+import copy
+
+if sys.version_info > (3, 0):
+    from .s3_service import S3Service
+else:
+    from s3_service import S3Service
+
+
+class Storage(object):
+    """
+    A Storage object is used by executors and other components to access underlying storage service
+    without exposing the the implementation details.
+    Currently we only support S3 as the underlying service.
+    We plan to support other services in the future.
+    """
+
+    def __init__(self, config):
+        self.storage_config = config
+        self.prefix = config['storage_prefix']
+        self.service = config['storage_service']
+        if config['storage_service'] == 's3':
+            self.service_handler = S3Service(config['s3'])
+        else:
+            raise NotImplementedError(("Using {} as storage service is" +
+                                       "not supported yet").format(config['storage_service']))
+
+    def get_storage_config(self):
+        """
+        Retrieves the configuration of this storage handler.
+        :return: storage configuration
+        """
+        return self.storage_config
+
+    def get_storage_info(self):
+        """
+        Get the information of underlying storage service.
+        :return:
+        """
+        info = dict()
+        info['service'] = self.service
+        info['location'] = self.service_handler.get_storage_location()
+        return info
+
+    def put_object(self, key, data):
+        """
+        Put an object into storage.
+        :param key: object key
+        :param data: object data
+        :return: None
+        """
+        return self.service_handler.put_object(key, data)
+
+    def get_object(self, key):
+        """
+        Get an object with key.
+        :param key: object key
+        :return: data
+        """
+        return self.service_handler.get_object(key)
+
+    def create_keys(self, callset_id, call_id):
+        """
+        Create keys for data, output and status given callset and call IDs.
+        :param callset_id: callset's ID
+        :param call_id: call's ID
+        :return: data_key, output_key, status_key
+        """
+        data_key = os.path.join(self.prefix, callset_id, call_id, "data.pickle")
+        output_key = os.path.join(self.prefix, callset_id, call_id, "output.pickle")
+        status_key = os.path.join(self.prefix, callset_id, call_id, "status.json")
+        return data_key, output_key, status_key
+
+    def create_func_key(self, callset_id):
```

Comment:
This is related to the previous comment -- I feel like we should only have the abstract methods in this class so it is very clear what needs to be implemented for a new implementation ?